### PR TITLE
Add trace activity-buffering logic when DeliveryMode is "expectReplies"

### LINKED ISSUE #2732: Trace activities not filtered on DeliveryMode.ExpectReplies
## Versions
What package version of the SDK are you using. 4.9.3
What nodejs version are you using 12.16.1
What browser version are you using -
What os are you using -

## Describe the bug
When using `DeliveryMode.ExpectReplies` to send an activity using directline (e.g. also used to communicate with a skill), the response will not run through the same filter as with a non-expected replies activity.
Instead of filtering `Trace` activities, every type of activity will go into the `bufferedResponses` and be sent to the client.

## To Reproduce
1. Running a simple JS sample that sends a response.
2. Send an activity using `DeliveryMode` = `ExpectReplies`.
3. Observe `BotFrameworkAdapter.sendActivities` never being called which results in the funny comment [`// Just eat activity`](https://github.com/microsoft/botbuilder-js/blob/54e4c2ded75436c00fe182dcb4c0c077f45741bc/libraries/botbuilder/src/botFrameworkAdapter.ts#L980) and more importantly the result of it filtering traces never being executed. The issue starts in [`TurnContext.sendActivities`](https://github.com/microsoft/botbuilder-js/blob/31e2d7aabbdf99b91323621cf6f1077e03228410/libraries/botbuilder-core/src/turnContext.ts#L468) which will not call the adapter's `sendActivities` but the adapter continuing the `processActivity` execution which will send the `TurnContext.bufferedReplyActivities`.

## Expected behavior
Same behaviour when calling `sendActivity/ies` for either type of `DeliveryMode`, which filters traces. Alternatively, the possibility to set an option to also filter trace activities for these scenarios.

## Screenshots
\-

## Additional context
\-

### PR DESCRIPTION
**Pull request by stevengum**
Fixes #2732

FYI @alopix, @johnataylor, @EricDahlvang 

## Description
When DeliveryMode is `"expectReplies"`, outgoing trace activities are not routed through the TurnContext's BotAdapter. This resulted in trace activities being sent to the channel. This PR duplicates the trace activity logic in BotFrameworkAdapter.sendActivities() and places it in TurnContext.sendActivities().

#### BotFrameworkAdapter.sendActivities():
https://github.com/microsoft/botbuilder-js/blob/015ebb7c24a7dc50320fe34f13e84e6b5a4fcdbc/libraries/botbuilder/src/botFrameworkAdapter.ts#L979-L981

There are a couple of ways to address issue #2732, including via BotFrameworkAdapter.processActivity() or SkillDialog/BeginSkill. This current implementation at https://github.com/microsoft/botbuilder-js/commit/02aff60a70ac567b48415cccb0f6ae924e0077e3 is perhaps the most straightforward way of fixing it, but may have unintended consequences for channels with custom clients/3rd-party channels that do support trace activities.

Additionally, having the trace filtering logic in BotFrameworkAdapter.processActivity() means the "restriction" is at the same level (an integration library, `botbuilder`) as opposed to having the logic in a base concept library (`botbuilder-core`).

## Specific Changes
  - Move away from magic strings in `BotFrameworkAdapter.sendActivities()`
  - Duplicate `trace` activity logic in `TurnContext.sendActivities()`
  - Add unit tests for `TurnContext` changes & minor test cleanup

## Testing
Added unit tests

### CODE REVIEW

**Comment by stevengum** [comment]
@Gabog, right now I strongly believe we should add expectReplies bufferedReplies-parsing logic at the BotFrameworkAdapter level. We should also setup the Adapter to have a hook for filtering out certain activities in bufferedReplies and configure tracing to be an explicit opt-in.

**Comment by joshgummersall** [comment]
@stevengum do we intend to land this change? I see it's linked to an open R11 issue.

**Comment by stevengum** [comment]
Yes, this will probably land early next week. I'll sync with @gabog and @tracyboehrer to verify resource availability for porting it to .NET and Python

**Review by joshgummersall — APPROVED** [review]
This seems okay to me, though I don't think I fully understand the potential issues you mentioned in the description.

File: libraries/botbuilder/src/botFrameworkAdapter.ts [hunk 1]
```diff
@@ -1195,8 +1196,17 @@ export class BotFrameworkAdapter
...
                 // Handle "expectReplies" scenarios where all the activities have been buffered and sent back at once
                 // in an invoke response.
-                const expectedReplies: ExpectedReplies = { activities: context.bufferedReplyActivities as Activity[] };
-                body = expectedReplies;
+                let activities = context.bufferedReplyActivities as Activity[];
+                
+                // If the channel is not the emulator, do not send trace activities.
+                // Fixes: https://github.com/microsoft/botbuilder-js/issues/2732
+                if (request.channelId !== Channels.Emulator) {
+                    activities = context.bufferedReplyActivities.reduce(
```
Comment (joshgummersall) [inline libraries/botbuilder/src/botFrameworkAdapter.ts:1204]: Seems like `filter` would do this in a more concise way: `activities = activities.filter(a => a.type === ActivityTypes.Trace)`

**Review by joshgummersall — APPROVED** [review]
Tiny suggestion, otherwise LGTM.

File: libraries/botbuilder/src/botFrameworkAdapter.ts (hunk 1 above)
Comment (stevengum) [reply libraries/botbuilder/src/botFrameworkAdapter.ts:1204]: Good call, I don't know why I led with `.reduce()`...

Comment (joshgummersall) [reply libraries/botbuilder/src/botFrameworkAdapter.ts:1204]: Reduce is a good one too 😄

**Review by joshgummersall — APPROVED** [review]

**Review by johnataylor — APPROVED** [review]

**Pull request merged**